packages/dashboard/src/store/actions/updateAssetQuery/index.ts

```diff
@@ -0,0 +1,72 @@
+import { AssetQuery } from '@iot-app-kit/core';
+import { Action } from 'redux';
+
+import { AppKitWidget, Widget } from '../../../types';
+import { DashboardState } from '../../state';
+import { colorPalette } from '../../../util/colorPalette';
+
+type UpdateAssetQueryActionPayload = {
+  widget: AppKitWidget;
+  assetQuery: AssetQuery[];
+};
+
+export interface UpdateAssetQueryAction extends Action {
+  type: 'UPDATE_ASSET_QUERY';
+  payload: UpdateAssetQueryActionPayload;
+}
+
+const assignDefaultRefId = (queries: AssetQuery[]) => {
+  return queries.map<AssetQuery>(({ properties, ...others }) => ({
+    ...others,
+    properties: properties.map((propertyQuery) => ({
+      ...propertyQuery,
+      refId: propertyQuery.refId || propertyQuery.propertyId,
+    })),
+  }));
+};
+export const onUpdateAssetQueryAction = (payload: UpdateAssetQueryActionPayload): UpdateAssetQueryAction => {
+  const { widget, assetQuery } = payload;
+  const styleSettings = (widget as AppKitWidget).styleSettings || {};
+
+  const assetQueryWithRefId = assignDefaultRefId(assetQuery);
+  // assign default color
+  assetQueryWithRefId
+    .flatMap(({ properties }) => properties)
+    .forEach(({ refId }, index) => {
+      if (refId && !styleSettings[refId]) {
+        styleSettings[refId] = {
+          color: colorPalette[index % colorPalette.length],
+        };
+      }
+    });
+
+  return {
+    type: 'UPDATE_ASSET_QUERY',
+    payload: { widget: { ...widget, styleSettings }, assetQuery: assetQueryWithRefId },
+  };
+};
+
+export const updateAssetQuery = (state: DashboardState, action: UpdateAssetQueryAction): DashboardState => {
+  const findAndUpdateWidgetById = (widgets: Widget[]) =>
+    widgets.map((w) => {
+      if (w.id === action.payload.widget.id) {
+        return {
+          ...action.payload.widget,
+          assets: action.payload.assetQuery,
```

Comment:
Will this delete asset(s) previously attached to the widget and replace them with new asset(s) from the payload? Do we always want to do that, or are there circumstances in which we might want to add assets to a widget with some assets already? I'm thinking particularly of dragging multiple assets/properties from the resource explorer onto a single chart, do we want to 'add' each to the chart, or replace what's already there?

Reply:
(Not necessarily saying the way you're doing it here is wrong, totally fine if this is a first step and that's a feature to be added later)

Reply:
UpdateAssetQuery simply replaces widget's assets list with the new payload. We might want to have 2 extra actions like RemoveAssetQuery and AddAssetQuery to handle removing and adding assets. Or we need to handle them when creating the payload.